### PR TITLE
Fix list streams crashing if parent object was deleted.

### LINKED ISSUE #7242: java.lang.IllegalStateException: Only valid, managed RealmLists can be frozen
#### Goal
I was using Kotlin Flow with the default InternalFlowFactory(returnFrozenObjects = true). I have a coroutine launched by the viewLifeCycleScope of a fragment. Inside that lifecycle i have a realm query and create a flow from it that updates the UI. I would suspect i can safely delete the object where the flow is attached to.

#### Actual Results
Upon deletion the app crashes. There is no reference to my code in the logs. The crash goes away if i set InternalFlowFactory(returnFrozenObjects = false), however this is not what i want by default.

```
JNI DETECTED ERROR IN APPLICATION: JNI NewLocalRef called with pending exception java.lang.IllegalStateException: Only valid, managed RealmLists can be frozen.
    java_vm_ext.cc:577]   at io.realm.RealmList io.realm.RealmList.freeze() (RealmList.java:168)
    java_vm_ext.cc:577]   at void io.realm.internal.coroutines.InternalFlowFactory$from$5$listener$1.onChange(io.realm.RealmList) (InternalFlowFactory.kt:324)
    java_vm_ext.cc:577]   at void io.realm.internal.coroutines.InternalFlowFactory$from$5$listener$1.onChange(java.lang.Object) (InternalFlowFactory.kt:34)
    java_vm_ext.cc:577]   at void io.realm.internal.ObservableCollection$RealmChangeListenerWrapper.onChange(java.lang.Object, io.realm.OrderedCollectionChangeSet) (ObservableCollection.java:39)
    java_vm_ext.cc:577]   at void io.realm.internal.ObservableCollection$CollectionObserverPair.onChange(java.lang.Object, io.realm.internal.OsCollectionChangeSet) (ObservableCollection.java:20)
    java_vm_ext.cc:577]   at void io.realm.internal.ObservableCollection$Callback.onCalled(io.realm.internal.ObservableCollection$CollectionObserverPair, java.lang.Object) (ObservableCollection.java:64)
    java_vm_ext.cc:577]   at void io.realm.internal.ObservableCollection$Callback.onCalled(io.realm.internal.ObserverPairList$ObserverPair, java.lang.Object) (ObservableCollection.java:54)
    java_vm_ext.cc:577]   at void io.realm.internal.ObserverPairList.foreach(io.realm.internal.ObserverPairList$Callback) (ObserverPairList.java:109)
    java_vm_ext.cc:577]   at void io.realm.internal.OsList.notifyChangeListeners(long) (OsList.java:314)
    java_vm_ext.cc:577]   at void android.os.MessageQueue.nativePollOnce(long, int) (MessageQueue.java:-2)
    java_vm_ext.cc:577]   at android.os.Message android.os.MessageQueue.next() (MessageQueue.java:335)
    java_vm_ext.cc:577]   at void android.os.Looper.loop() (Looper.java:183)
    java_vm_ext.cc:577]   at void android.app.ActivityThread.main(java.lang.String[]) (ActivityThread.java:7656)
    java_vm_ext.cc:577]   at java.lang.Object java.lang.reflect.Method.invoke(java.lang.Object, java.lang.Object[]) (Method.java:-2)
    java_vm_ext.cc:577]   at void com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run() (RuntimeInit.java:592)
    java_vm_ext.cc:577]   at void com.android.internal.os.ZygoteInit.main(java.lang.String[]) (ZygoteInit.java:947)
    java_vm_ext.cc:577] 
    java_vm_ext.cc:577]     in call to NewLocalRef
    java_vm_ext.cc:577]     from void android.os.MessageQueue.nativePollOnce(long, int)
```

Shouldn't the InternalFlowFactory check whether the object is still valid?
I tried with the RxJava version and also get the same error.

#### Version of Realm and tooling

Realm version(s): 10.2.0

Realm Sync feature enabled: No

Android Studio version: 4.1.1

Android Build Tools version: 29.0.3

Gradle version: 4.1.1

Which Android version and device(s): Pixel 3 Android 11

### PR DESCRIPTION
**Pull request by cmelchior**
Finally managed to get Android Studio working again 🙄 

Closes #7242 

This PR fixes a bug in our RxJava/Corutine implementation that didn't correctly handle streams from lists when the parent object holding the list was deleted. 

How to handle it is a bit up for debate though. I could see 4 choices: 

* Either emit the invalid collection. In which case people would have to guard for it. But it would return false for frozen unlike everything else.
* Return a `null` collection and then close the stream. This would introduce null checks everywhere in the Kotlin API
* Throw an exception in the stream.
* Do not emit anything and just close the stream gracefully. 

I opted for the last option as deleting parents seems like a valid use case, and completing the stream does indicate that no further changes will be coming. Alternatively, I was also leaning towards 2, but it does make consuming the API a bit annoying for the common case.

Any thoughts on this?

I also exposed our `BlockingLooperThread` as a Dispatcher. I think it can be useful to clean up some of the other tests (In another PR).

### CODE REVIEW

File: realm/kotlin-extensions/src/androidTest/kotlin/io/realm/CoroutinesTests.kt [hunk 1]
```diff
@@ -924,6 +928,160 @@ class CoroutinesTests {
...
+        scope.launch {
+            // We should only emit valid lists. If the parent of the list is invalidated
+            // it should close the stream gracefully resulting in onComplete being called.
+            list.toFlow()
+                    .onEach { assertTrue(it.isValid) }
+                    .onCompletion {
+                        realm.close()
+                        looperThread.testComplete()
+                    }.collect {
+                        collectingStarted.set(true)
```
Comment (rorbech) [inline realm/kotlin-extensions/src/androidTest/kotlin/io/realm/CoroutinesTests.kt:952]: Maybe not required for now, but I guess you could communicate through a channel now that we are in a coroutine context. That would make it cleaner when waiting for the result. Good old CSP.

**Review by rorbech — APPROVED** [review]
Looks good to me. I think closing the stream gracefully sound reasonable, but you should probably also have @edualonso's word on that too.

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,3 +1,19 @@
+## 10.2.1 (YYYY-MM-DD)
+
+### Enhancements
+* None.
+
+### Fixes
+* RxJava Flowables/Observables and Coroutine Flows would crash if they where created from a `RealmList` and the parent object holding the list was deleted. Now, the stream is disposed/closed instead. (Issue [#7242](https://github.com/realm/realm-java/issues/7242))
```
Comment (edualonso) [inline CHANGELOG.md:7]:
```suggestion
* RxJava Flowables/Observables and Coroutine Flows would crash if they were created from a `RealmList` and the parent object holding the list was deleted. Now, the stream is disposed/closed instead. (Issue [#7242](https://github.com/realm/realm-java/issues/7242)) 
```

File: realm/realm-library/src/testUtils/kotlin/io/realm/rule/BlockingLooperThread.kt [hunk 3]
```diff
@@ -222,6 +208,11 @@ class BlockingLooperThread {
         }
     }
 
+    fun asDispatcher(): CoroutineContext {
```
Comment (edualonso) [inline realm/realm-library/src/testUtils/kotlin/io/realm/rule/BlockingLooperThread.kt:211]: 👏

**Review by edualonso — APPROVED** [review]
Looks good!

**Pull request merged**